File: client-ios/SerenadaCore/Sources/SerenadaConfig.swift
import Foundation
import SerenadaBroadcastExtensionSupport

/// Signaling transport type.
public enum SerenadaTransport: String, Equatable, Sendable {
    case ws
    case sse
}

/// How the SDK captures screen-share video during a call.
///
/// Screen sharing is runtime-configured, not a compile-time variant: both
/// capture paths are always built and the SDK selects one from this value. The
/// default is `.disabled`, so a host that has not provisioned a broadcast
/// extension never accidentally targets another app's IPC channels.
public enum ScreenShareMode: Equatable, Sendable {
    /// Screen sharing is unavailable: `screenShareExtensionBundleId` is `nil` and
    /// the share control stays hidden. Default — hosts opt in explicitly.
    case disabled
    /// System-wide capture via a Broadcast Upload Extension, delivered over the
    /// app-group shared memory described by `BroadcastIPCConfig`. Captures the
    /// whole device screen and survives backgrounding.
    case broadcast(BroadcastIPCConfig)
    /// In-app ReplayKit capture (`RPScreenRecorder`): only this app's own content,
    /// foreground-only. SDK/reference scope; not for hosts needing full-device
    /// sharing.
    case inAppOnly
}

/// Default preference order for camera modes when `SerenadaConfig.cameraModes` is `nil`.
public let defaultCameraModes: [LocalCameraMode] = [.selfie, .world, .composite]

/// SDK configuration.
///
/// `SerenadaConfig` is marked `@unchecked Sendable` because a custom
/// `SignalingProvider` is class-bound and cannot be proven sendable by the
/// compiler. Callers are responsible for keeping provider implementations
/// thread-safe according to the provider contract.
public struct SerenadaConfig: Equatable, @unchecked Sendable {
    /// Server host or origin (e.g. "serenada.app" or "localhost:8080").
    public let serverHost: String?
    /// Custom signaling provider. Provide exactly one of `serverHost` or `signalingProvider`.
    public let signalingProvider: SignalingProvider?
    /// Whether audio is enabled when joining a call. Defaults to `true`.
    public var defaultAudioEnabled: Bool
    /// Whether video is enabled when joining a call. Defaults to `true`. When this is `false`,
    /// camera permission is not required for the initial join even when camera modes are
    /// available; the SDK requests camera access lazily if the user enables video later.
    public var defaultVideoEnabled: Bool
    /// Whether this call can negotiate any video media. Set to `false` for strict
    /// audio-only calls such as PSTN: camera capture, screen sharing, and remote
    /// video are all disabled. Defaults to `true`.
    public var videoMediaEnabled: Bool
    /// Static capability gate for the independent content (screen-share) video
    /// stream. When `true`, this client advertises `independentContentVideo` at
    /// `join` and (in later phases) negotiates a separate content transceiver.
    /// Immutable per session. Defaults to `false` until the platform's media
    /// engine and UI/API surface ship — a client with the flag off behaves
    /// exactly like today, including the legacy `cameraMode=screenShare` path.
    public var enableIndependentContentVideo: Bool
    /// How screen-share video is captured. Defaults to `.disabled`; set
    /// `.broadcast(...)` with the host's app group + extension bundle ID to enable
    /// full-device sharing, or `.inAppOnly` for in-app ReplayKit capture.
    public var screenShareMode: ScreenShareMode
    /// Camera modes available in the call UI, in preference order. The first
    /// entry is the initial mode. When only one mode is listed the flip-camera
    /// control is hidden; an empty array disables camera capture (the video
    /// toggle is hidden and the camera is never requested). Remote video and
    /// screen sharing remain available unless `videoMediaEnabled` is `false`.
    /// Modes unsupported on the current device are silently dropped (`.composite`
    /// is dropped on devices without multi-cam). `.screenShare` is always ignored
    /// — screen sharing is controlled separately. Defaults to `[.selfie, .world, .composite]`.
    public var cameraModes: [LocalCameraMode]?
    /// When `true`, defer the initial-negotiation offer-timeout/ICE-restart while the host peer
    /// awaits its first answer. Use for app-owned calls whose answer is gated on a remote action
    /// that may take longer than the offer timeout, such as PSTN pickup. Defaults to `false`.
    public var deferInitialAnswer: Bool
    /// Preferred signaling transports in priority order. Defaults to `[.ws, .sse]`.
    public var transports: [SerenadaTransport]
    /// Whether the proximity sensor is used to switch audio to the earpiece and pause video.
    /// Defaults to `false`.
    public var proximityMonitoringEnabled: Bool
    /// Custom audio coordinator. If `nil`, the SDK uses its internal default coordinator.
    public var audioCoordinator: SerenadaAudioCoordinator?
    /// Audio policy passed to the coordinator when a call session activates.
    public var audioIntent: AudioIntent

    public init(
        serverHost: String? = nil,
        signalingProvider: SignalingProvider? = nil,
        defaultAudioEnabled: Bool = true,
        defaultVideoEnabled: Bool = true,
        videoMediaEnabled: Bool = true,
        enableIndependentContentVideo: Bool = false,
        screenShareMode: ScreenShareMode = .disabled,
        cameraModes: [LocalCameraMode]? = nil,
        deferInitialAnswer: Bool = false,
        transports: [SerenadaTransport] = [.ws, .sse],
        proximityMonitoringEnabled: Bool = false,
        audioCoordinator: SerenadaAudioCoordinator? = nil,
        audioIntent: AudioIntent = AudioIntent()
    ) {
        self.serverHost = serverHost
        self.signalingProvider = signalingProvider
        self.defaultAudioEnabled = defaultAudioEnabled
        self.defaultVideoEnabled = defaultVideoEnabled
        self.videoMediaEnabled = videoMediaEnabled
        self.enableIndependentContentVideo = enableIndependentContentVideo
        self.screenShareMode = screenShareMode
        self.cameraModes = cameraModes
        self.deferInitialAnswer = deferInitialAnswer
        self.transports = transports
        self.proximityMonitoringEnabled = proximityMonitoringEnabled
        self.audioCoordinator = audioCoordinator
        self.audioIntent = audioIntent
    }

    public static func == (lhs: SerenadaConfig, rhs: SerenadaConfig) -> Bool {
        lhs.serverHost == rhs.serverHost
            && lhs.defaultAudioEnabled == rhs.defaultAudioEnabled
            && lhs.defaultVideoEnabled == rhs.defaultVideoEnabled
            && lhs.videoMediaEnabled == rhs.videoMediaEnabled
            && lhs.enableIndependentContentVideo == rhs.enableIndependentContentVideo
            && lhs.screenShareMode == rhs.screenShareMode
            && lhs.cameraModes == rhs.cameraModes
            && lhs.deferInitialAnswer == rhs.deferInitialAnswer
            && lhs.transports == rhs.transports
            && lhs.proximityMonitoringEnabled == rhs.proximityMonitoringEnabled
            && haveSameProvider(lhs.signalingProvider, rhs.signalingProvider)
            && lhs.audioCoordinator === rhs.audioCoordinator
            && lhs.audioIntent == rhs.audioIntent
    }
}

internal struct ResolvedSerenadaConfig {
    let serverHost: String?
    let signalingProvider: SignalingProvider?
}

internal let SUPPORTED_SIGNALING_PROVIDER_VERSION = 1

internal func resolveSerenadaConfig(_ config: SerenadaConfig) throws -> ResolvedSerenadaConfig {
    let serverHost = config.serverHost?
        .trimmingCharacters(in: .whitespacesAndNewlines)
        .nilIfEmpty
    let signalingProvider = config.signalingProvider

    guard (serverHost == nil) != (signalingProvider == nil) else {
        throw APIError.invalidResponse("Provide exactly one of serverHost or signalingProvider")
    }
    if let signalingProvider, signalingProvider.version != SUPPORTED_SIGNALING_PROVIDER_VERSION {
        throw APIError.invalidResponse("Unsupported signalingProvider version: \(signalingProvider.version)")
    }

    return ResolvedSerenadaConfig(
        serverHost: serverHost,
        signalingProvider: signalingProvider
    )
}

internal func requireServerHost(_ config: SerenadaConfig) throws -> String {
    let resolved = try resolveSerenadaConfig(config)
    guard let serverHost = resolved.serverHost else {
        throw APIError.invalidResponse("requires serverHost")
    }
    return serverHost
}

private func haveSameProvider(_ lhs: SignalingProvider?, _ rhs: SignalingProvider?) -> Bool {
    switch (lhs, rhs) {
    case (nil, nil):
        true
    case let (lhs?, rhs?):
        ObjectIdentifier(lhs) == ObjectIdentifier(rhs)
    default:
        false
    }
}

File: client-android/serenada-core/src/main/java/app/serenada/core/SerenadaConfig.kt
package app.serenada.core

import app.serenada.core.call.LocalCameraMode
import app.serenada.core.call.SerenadaAudioCoordinator
import app.serenada.core.call.AudioIntent

/** Default preference order for [SerenadaConfig.cameraModes] when left null. */
val DEFAULT_CAMERA_MODES: List<LocalCameraMode> = listOf(
    LocalCameraMode.SELFIE,
    LocalCameraMode.WORLD,
    LocalCameraMode.COMPOSITE,
)

/**
 * Configuration for the Serenada SDK.
 */
data class SerenadaConfig(
    /** Server host or origin (e.g. "serenada.app" or "http://localhost:8080"). */
    val serverHost: String? = null,
    /**
     * Custom SINGLE-SESSION signaling provider. Provide exactly one of `serverHost`,
     * `signalingProvider`, or `multiSessionSignalingProvider`. A v1 provider is bound
     * by at most one live session at a time; a second CONCURRENT session that would
     * bind it fails with a typed error ([CallError.ProviderUnavailable]). For multiple
     * concurrent sessions (multi-call), use [multiSessionSignalingProvider].
     */
    val signalingProvider: SignalingProvider? = null,
    /**
     * Custom MULTI-SESSION signaling provider (contract §"Custom provider"). Vends one
     * signaling channel per session so concurrent sessions never share a listener or
     * cross-wire their rooms. Provide exactly one of `serverHost`, `signalingProvider`,
     * or `multiSessionSignalingProvider`.
     */
    val multiSessionSignalingProvider: MultiSessionSignalingProvider? = null,
    /** Whether audio starts enabled (default true). */
    val defaultAudioEnabled: Boolean = true,
    /** Whether video starts enabled (default true). */
    val defaultVideoEnabled: Boolean = true,
    /**
     * Whether this call can negotiate any video media. Set false for strict
     * audio-only calls such as PSTN: camera capture, screen sharing, and
     * remote video are all disabled. Defaults to true.
     */
    val videoMediaEnabled: Boolean = true,
    /**
     * Static build capability: whether this client can negotiate, send,
     * receive, classify, expose, and render an independent content (screen
     * share) video stream. Advertised at `join` as
     * `capabilities.independentContentVideo`. Immutable per session.
     *
     * Defaults to false so external integrators keep the legacy single-video
     * screen-share behavior until their UI can render the content stream.
     */
    val enableIndependentContentVideo: Boolean = false,
    /**
     * Camera modes available in the call UI, in preference order. The first
     * entry is the initial mode. When only one mode is listed the flip-camera
     * control is hidden; an empty list disables camera capture (the video
     * toggle is hidden and the camera is never requested). Remote video and
     * screen sharing remain available unless [videoMediaEnabled] is false.
     * Modes unsupported on the current device are silently dropped
     * (`COMPOSITE` is dropped on devices without multi-cam). `SCREEN_SHARE` is
     * always ignored — screen sharing is controlled separately. Defaults to
     * `[SELFIE, WORLD, COMPOSITE]`.
     */
    val cameraModes: List<LocalCameraMode>? = null,
    /** Enable experimental HD video capture. */
    val isHdVideoExperimentalEnabled: Boolean = false,
    /**
     * When true, defer the initial-negotiation offer-timeout/ICE-restart while the host peer
     * awaits its FIRST answer. For app-owned calls whose answer is gated on a remote action that
     * may take much longer than the offer timeout (e.g. PSTN human pickup). Normal offer-timeout
     * behavior resumes after the first answer. Default false = unchanged for existing calls.
     */
    val deferInitialAnswer: Boolean = false,
    /** Preferred signaling transports in priority order (default: WS then SSE). */
    val transports: List<SerenadaTransport> = listOf(SerenadaTransport.WS, SerenadaTransport.SSE),
    /** Whether the proximity sensor is used to switch audio to the earpiece and pause video (default false). */
    val proximityMonitoringEnabled: Boolean = false,
    /** Custom audio coordinator. If null, the SDK uses its internal default coordinator. */
    val audioCoordinator: SerenadaAudioCoordinator? = null,
    /** Audio policy passed to the coordinator when a call session activates. */
    val audioIntent: AudioIntent = AudioIntent(),
)

/** Available signaling transport types. */
enum class SerenadaTransport {
    /** WebSocket transport. */
    WS,
    /** Server-Sent Events transport. */
    SSE,
}

internal data class ResolvedSerenadaConfig(
    val serverHost: String?,
    val signalingProvider: SignalingProvider?,
    val multiSessionSignalingProvider: MultiSessionSignalingProvider? = null,
)

internal const val SUPPORTED_SIGNALING_PROVIDER_VERSION = 1
internal const val MULTI_SESSION_SIGNALING_PROVIDER_VERSION = 2

internal fun resolveSerenadaConfig(config: SerenadaConfig): ResolvedSerenadaConfig {
    val serverHost = config.serverHost?.trim()?.takeIf { it.isNotEmpty() }
    val signalingProvider = config.signalingProvider
    val multiSessionSignalingProvider = config.multiSessionSignalingProvider
    require(signalingProvider == null || multiSessionSignalingProvider == null) {
        "Provide only one of signalingProvider or multiSessionSignalingProvider"
    }
    val hasProvider = signalingProvider != null || multiSessionSignalingProvider != null
    require((serverHost != null) != hasProvider) {
        // Message kept verbatim for backward compatibility with existing integrators.
        "Provide exactly one of serverHost or signalingProvider"
    }
    if (signalingProvider != null && signalingProvider.version != SUPPORTED_SIGNALING_PROVIDER_VERSION) {
        throw IllegalArgumentException("Unsupported signalingProvider version: ${signalingProvider.version}")
    }
    if (multiSessionSignalingProvider != null &&
        multiSessionSignalingProvider.version != MULTI_SESSION_SIGNALING_PROVIDER_VERSION
    ) {
        throw IllegalArgumentException(
            "Unsupported multiSessionSignalingProvider version: ${multiSessionSignalingProvider.version}",
        )
    }
    return ResolvedSerenadaConfig(
        serverHost = serverHost,
        signalingProvider = signalingProvider,
        multiSessionSignalingProvider = multiSessionSignalingProvider,
    )
}

internal fun requireServerHost(config: SerenadaConfig): String {
    return resolveSerenadaConfig(config).serverHost
        ?: throw IllegalStateException("requires serverHost")
}
